File: Problems Solutions/707. Design Linked List.py
# https://leetcode.com/problems/design-linked-list/description

from utils import list_to_linked_list, ListNode, is_same_linked_lists

class Node:
    def __init__(self, v, next=None):
        self.value = v
        self.next = next

class MyLinkedList:

    def __init__(self):
        self.head = None
        self.count = 0
    
    def __get_node_by_index(self, index) -> Node:
        node = self.head
        
        for _ in range(index):
            node = node.next
        
        return node        

    def get(self, index: int) -> int:
        if index not in range(0, self.count):
            return -1
        return self.__get_node_by_index(index).value

    def addAtHead(self, val: int) -> None:
        new_head = Node(val, self.head)
        self.head = new_head
        self.count += 1

    def addAtTail(self, val: int) -> None:
        if not self.head:
            self.head = Node(val)
        else:
            last_node = self.__get_node_by_index(self.count - 1)
            last_node.next = Node(val)
        self.count += 1

    def addAtIndex(self, index: int, val: int) -> None:
        if index not in range(0, self.count + 1):
            return

        if index == self.count:
            self.addAtTail(val)
        elif index == 0:
            self.addAtHead(val)
        else:
            node = self.__get_node_by_index(index - 1)
            next = node.next
            node.next = Node(val)
            node.next.next = next
            self.count += 1

    def deleteAtIndex(self, index: int) -> None:
        if index not in range(0, self.count):
            return
        
        if index == 0:
            self.head = self.head.next
        elif index == self.count - 1:
            self.__get_node_by_index(self.count - 2).next = None
        else:
            node = self.__get_node_by_index(index - 1)
            node.next = node.next.next
        
        self.count -= 1

File: Problems Solutions/go/707. Design Linked List.go
package main

import "fmt"

type Node struct {
	value int
	next  *Node
}

type MyLinkedList struct {
	head  *Node
	count int
}

func CreateFrom(items []int) MyLinkedList {
	var list = MyLinkedList{}
	for _, value := range(items){
		list.AddAtTail(value)
	}
	return list
}

func Constructor() MyLinkedList {
	var list = MyLinkedList{count: 0, head: nil}
	return list
}

func (l *MyLinkedList) Print() {
	var node = l.head
	fmt.Println(node.value)

	for node.next != nil {
		node = node.next
		fmt.Println(node.value)
	}
}

func (this *MyLinkedList) getNode(index int) *Node {
	var node = this.head
	for i := 0; i < index; i++ {
		node = node.next
	}
	return node
}

func (this *MyLinkedList) Get(index int) int {
	if index > this.count-1 {
		return -1
	}

	return this.getNode(index).value
}

func (this *MyLinkedList) AddAtHead(val int) {
	var node *Node = &Node{value: val}

	if this.head == nil {
		this.head = node
	} else {
		var oldHead = this.head
		this.head = node
		this.head.next = oldHead
	}

	this.count++
}

func (this *MyLinkedList) AddAtTail(val int) {
	var node *Node = &Node{value: val}

	if this.head == nil {
		this.head = node
	} else {
		this.getNode(this.count - 1).next = node
	}

	this.count++
}

func (this *MyLinkedList) AddAtIndex(index int, val int) {
	if index > this.count {
		return
	}

	if index == this.count {
		this.AddAtTail(val)
	} else if index == 0 {
		this.AddAtHead(val)
	} else {
		var prevNode = this.getNode(index - 1)
		var nextNode = prevNode.next
		prevNode.next = &Node{value: val}
		prevNode.next.next = nextNode
		this.count++
	}
}

func (this *MyLinkedList) DeleteAtIndex(index int) {
	if index > this.count-1 {
		return
	}

	if index == this.count-1 {
		this.getNode(this.count - 2).next = nil
	} else if index == 0 {
		this.head = this.head.next
	} else {
		var prevNode = this.getNode(index - 1)
		prevNode.next = prevNode.next.next
	}

	this.count--
}

func main() {
	var l = Constructor()
	l.AddAtHead(1)
	l.AddAtTail(3)
	l.AddAtIndex(1, 2)
	l.DeleteAtIndex(1)
	fmt.Println(l.Get(1), "\n")

	var nums = []int {1, 2, 3, 4, 5}
	l = CreateFrom(nums)
	l.Print()
}
